### PR TITLE
bbox_iou() stability and speed improvements

### LINKED ISSUE #2376: a bug about function bbox_iou()
## 🐛 Bug
In function bbox_iou() of general.py, if box1 is equal to box2, it should output 1.0, but we get NaN. And the default eps is too small for Half precision.

## To Reproduce (REQUIRED)

Input:
(1)
a = torch.tensor([[1.42969, 0.25635, 3.78125, 7.37109]], dtype=torch.float32)
b = torch.tensor([[1.42969, 0.25635, 3.78125, 7.37109]], dtype=torch.float32)
import math
giou = bbox_iou(a.T, b, x1y1x2y2=False, CIoU=True)  # giou(prediction, target)
print(giou)

it will output NaN, and if we change a line in bbox_iou(), ''alpha = v / ((1 + eps) - iou + v)'' ==> ''alpha = v / ((1 - iou + v)+eps)'',  we will get the right output 1.0;

(2)
a = torch.tensor([[1.42969, 0.25635, 3.78125, 7.37109]], device='cuda:0', dtype=torch.float16)
b = torch.tensor([[1.42969, 0.25635, 3.78125, 7.37109]], device='cuda:0', dtype=torch.float16)
import math
giou = bbox_iou(a.T, b, x1y1x2y2=False, CIoU=True)  # giou(prediction, target)
print(giou)

it will output NaN, and if we use 3e-8 as the default eps rather than 1e-9,  we will get the right output 1.0;

### PR DESCRIPTION
**Pull request by glenn-jocher**
Partial fix for https://github.com/ultralytics/yolov5/issues/2376

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved robustness in bounding box intersection-over-union (IoU) calculations.

### 📊 Key Changes
- Adjusted the epsilon value in the `bbox_iou` function from `1e-9` to `1e-7`.
- Modified the `alpha` calculation formula within the CIoU condition.

### 🎯 Purpose & Impact
- **Enhanced Numerical Stability**: By tweaking the epsilon value, calculations become more stable, reducing the likelihood of division by zero or near-zero scenarios that can lead to numerical errors.
- **Increased Precision**: The new `alpha` calculation ensures higher precision, potentially improving the performance of bounding box regressions during training and inference.
- **Better Accuracy**: These adjustments could lead to more accurate object detection results, positively affecting tasks using the YOLOv5 model. 🎯✨

Overall, users can expect enhanced model stability and possibly improved detection metrics.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
Test code:

```python
import math
import torch


def bbox_iou(box1, box2, x1y1x2y2=True, GIoU=False, DIoU=False, CIoU=False, eps=1e-7):
    # Returns the IoU of box1 to box2. box1 is 4, box2 is nx4
    box2 = box2.T

    # Get the coordinates of bounding boxes
    if x1y1x2y2:  # x1, y1, x2, y2 = box1
        b1_x1, b1_y1, b1_x2, b1_y2 = box1[0], box1[1], box1[2], box1[3]
        b2_x1, b2_y1, b2_x2, b2_y2 = box2[0], box2[1], box2[2], box2[3]
    else:  # transform from xywh to xyxy
        b1_x1, b1_x2 = box1[0] - box1[2] / 2, box1[0] + box1[2] / 2
        b1_y1, b1_y2 = box1[1] - box1[3] / 2, box1[1] + box1[3] / 2
        b2_x1, b2_x2 = box2[0] - box2[2] / 2, box2[0] + box2[2] / 2
        b2_y1, b2_y2 = box2[1] - box2[3] / 2, box2[1] + box2[3] / 2

    # Intersection area
    inter = (torch.min(b1_x2, b2_x2) - torch.max(b1_x1, b2_x1)).clamp(0) * \
            (torch.min(b1_y2, b2_y2) - torch.max(b1_y1, b2_y1)).clamp(0)

    # Union Area
    w1, h1 = b1_x2 - b1_x1, b1_y2 - b1_y1 + eps
    w2, h2 = b2_x2 - b2_x1, b2_y2 - b2_y1 + eps
    union = w1 * h1 + w2 * h2 - inter + eps

    iou = inter / union
    if GIoU or DIoU or CIoU:
        cw = torch.max(b1_x2, b2_x2) - torch.min(b1_x1, b2_x1)  # convex (smallest enclosing box) width
        ch = torch.max(b1_y2, b2_y2) - torch.min(b1_y1, b2_y1)  # convex height
        if CIoU or DIoU:  # Distance or Complete IoU https://arxiv.org/abs/1911.08287v1
            c2 = cw ** 2 + ch ** 2 + eps  # convex diagonal squared
            rho2 = ((b2_x1 + b2_x2 - b1_x1 - b1_x2) ** 2 +
                    (b2_y1 + b2_y2 - b1_y1 - b1_y2) ** 2) / 4  # center distance squared
            if DIoU:
                return iou - rho2 / c2  # DIoU
            elif CIoU:  # https://github.com/Zzh-tju/DIoU-SSD-pytorch/blob/master/utils/box/box_utils.py#L47
                v = (4 / math.pi ** 2) * torch.pow(torch.atan(w2 / h2) - torch.atan(w1 / h1), 2)
                with torch.no_grad():
                    alpha = v / (v - iou + (1 + eps))
                return iou - (rho2 / c2 + v * alpha)  # CIoU
        else:  # GIoU https://arxiv.org/pdf/1902.09630.pdf
            c_area = cw * ch + eps  # convex area
            return iou - (c_area - union) / c_area  # GIoU
    else:
        return iou  # IoU


a = torch.tensor([[1.42969, 0.25635, 3.78125, 7.37109]], dtype=torch.float32)
b = torch.tensor([[1.42969, 0.25635, 3.78125, 7.37109]], dtype=torch.float32)
print(bbox_iou(a.T, b, x1y1x2y2=False, CIoU=True, eps=1e-7))
```

**Comment by John-chen521** [comment]
> Test code:
> 
> ```python
> import math
> import torch
> 
> 
> def bbox_iou(box1, box2, x1y1x2y2=True, GIoU=False, DIoU=False, CIoU=False, eps=1e-7):
>     # Returns the IoU of box1 to box2. box1 is 4, box2 is nx4
>     box2 = box2.T
> 
>     # Get the coordinates of bounding boxes
>     if x1y1x2y2:  # x1, y1, x2, y2 = box1
>         b1_x1, b1_y1, b1_x2, b1_y2 = box1[0], box1[1], box1[2], box1[3]
>         b2_x1, b2_y1, b2_x2, b2_y2 = box2[0], box2[1], box2[2], box2[3]
>     else:  # transform from xywh to xyxy
>         b1_x1, b1_x2 = box1[0] - box1[2] / 2, box1[0] + box1[2] / 2
>         b1_y1, b1_y2 = box1[1] - box1[3] / 2, box1[1] + box1[3] / 2
>         b2_x1, b2_x2 = box2[0] - box2[2] / 2, box2[0] + box2[2] / 2
>         b2_y1, b2_y2 = box2[1] - box2[3] / 2, box2[1] + box2[3] / 2
> 
>     # Intersection area
>     inter = (torch.min(b1_x2, b2_x2) - torch.max(b1_x1, b2_x1)).clamp(0) * \
>             (torch.min(b1_y2, b2_y2) - torch.max(b1_y1, b2_y1)).clamp(0)
> 
>     # Union Area
>     w1, h1 = b1_x2 - b1_x1, b1_y2 - b1_y1 + eps
>     w2, h2 = b2_x2 - b2_x1, b2_y2 - b2_y1 + eps
>     union = w1 * h1 + w2 * h2 - inter + eps
> 
>     iou = inter / union
>     if GIoU or DIoU or CIoU:
>         cw = torch.max(b1_x2, b2_x2) - torch.min(b1_x1, b2_x1)  # convex (smallest enclosing box) width
>         ch = torch.max(b1_y2, b2_y2) - torch.min(b1_y1, b2_y1)  # convex height
>         if CIoU or DIoU:  # Distance or Complete IoU https://arxiv.org/abs/1911.08287v1
>             c2 = cw ** 2 + ch ** 2 + eps  # convex diagonal squared
>             rho2 = ((b2_x1 + b2_x2 - b1_x1 - b1_x2) ** 2 +
>                     (b2_y1 + b2_y2 - b1_y1 - b1_y2) ** 2) / 4  # center distance squared
>             if DIoU:
>                 return iou - rho2 / c2  # DIoU
>             elif CIoU:  # https://github.com/Zzh-tju/DIoU-SSD-pytorch/blob/master/utils/box/box_utils.py#L47
>                 v = (4 / math.pi ** 2) * torch.pow(torch.atan(w2 / h2) - torch.atan(w1 / h1), 2)
>                 with torch.no_grad():
>                     alpha = v / (v - iou + (1 + eps))
>                 return iou - (rho2 / c2 + v * alpha)  # CIoU
>         else:  # GIoU https://arxiv.org/pdf/1902.09630.pdf
>             c_area = cw * ch + eps  # convex area
>             return iou - (c_area - union) / c_area  # GIoU
>     else:
>         return iou  # IoU
> 
> 
> a = torch.tensor([[1.42969, 0.25635, 3.78125, 7.37109]], dtype=torch.float32)
> b = torch.tensor([[1.42969, 0.25635, 3.78125, 7.37109]], dtype=torch.float32)
> print(bbox_iou(a.T, b, x1y1x2y2=False, CIoU=True, eps=1e-7))
> ```

hello How do you get the four values of your tensor？

**Comment by John-chen521** [comment]
Thanks，this code really useful！
